### PR TITLE
Fixed up Self-Ask Example

### PR DESCRIPTION
**Pull request by ivanleomk**
Added an example for self-ask

----

| :rocket: | This description was created by [Ellipsis](https://www.ellipsis.dev) for commit ae9fac5b218fbcd0f3803bef777c94cdc2ce274b  | 
|--------|--------|

### Summary:
Added detailed examples and documentation for the Self-Ask technique, updated navigation, provided alternative phrases for generating a thread of thought in prompts, and fixed a typo in `docs/concepts/patching.md`.

**Key points**:
- Added `docs/prompting/index.md` listing 58 prompting techniques, including Self-Ask.
- Updated `docs/prompting/zero_shot/self_ask.md` with a detailed Self-Ask example using the Instructor library.
- Added `docs/prompting/thought_generation/chain_of_thought_zero_shot/thread_of_thought.md` with a detailed Self-Ask example.
- Updated `mkdocs.yml` to include new Self-Ask documentation in navigation.
- Provided alternative phrases for generating a thread of thought in prompts.
- Fixed a typo in `docs/concepts/patching.md`.

----
Generated with :heart: by [ellipsis.dev](https://www.ellipsis.dev)

### CODE REVIEW

File: docs/prompting/thought_generation/chain_of_thought_zero_shot/thread_of_thought.md [hunk 1]
```diff
@@ -1,7 +1,128 @@
...
+        ...,
+        description="""This is a list that should contain a single
+        explanation for a relevant source""",
+    )
+    answer: str = Field(
+        ..., description="""This is the answer to the user's question"""
+    )
+
+
+def get_response():
```
Comment (jxnl) [inline docs/prompting/thought_generation/chain_of_thought_zero_shot/thread_of_thought.md:37]: name the function better, and make the sources and questions parameters so other people can actually use the code.

**Review by jxnl — CHANGES_REQUESTED** [review]
Again, if we copy-pasted this code, it wouldn't be useful.

Name the functions better and make sure things are actually parameters so other people can copy-paste this code and use it for themselves.

File: docs/prompting/thought_generation/chain_of_thought_zero_shot/tab_cot.md [hunk 2]
```diff
@@ -1,7 +1,89 @@
...
+import instructor
+from openai import OpenAI
+from pydantic import BaseModel, Field
+
+
+client = instructor.from_openai(OpenAI())
+
+
+class ReasoningStep(BaseModel):
+    step: int = Field(..., description="This is the step number")
```
Comment (jxnl) [inline docs/prompting/thought_generation/chain_of_thought_zero_shot/tab_cot.md:19]: dont use "This is a" just "The step Number"

File: docs/prompting/thought_generation/chain_of_thought_zero_shot/thread_of_thought.md [hunk 3]
```diff
@@ -1,7 +1,128 @@
...
+
+We can implement Thread Of Thought using the following template.
+
+!!! example "Thread Of Thought template"
+
+    **[ Input Prompt ]**
+
+    Proceed through the context systematically, zeroing in on areas that could provide the answers we’re seeking
+
+We can implement this in Instructor as seen below
```
Comment (jxnl) [inline docs/prompting/thought_generation/chain_of_thought_zero_shot/thread_of_thought.md:15]:
```suggestion
We can implement this using `instructor` as seen below
```

File: docs/prompting/thought_generation/chain_of_thought_zero_shot/tab_cot.md [hunk 4]
```diff
@@ -1,7 +1,90 @@
...
+
+
+def generate_structured_reasoning_response(query: str, context: str):
+    response = client.chat.completions.create(
+        model="gpt-4o",
+        response_model=Response,
+        messages=[
+            {
+                "role": "system",
+                "content": f"""
```
Comment (jxnl) [inline docs/prompting/thought_generation/chain_of_thought_zero_shot/tab_cot.md:51]: try lets add some more prompts in this string

File: docs/prompting/thought_generation/chain_of_thought_zero_shot/thread_of_thought.md [hunk 5]
```diff
@@ -1,7 +1,129 @@
...
+                    You are an expert Question Answerer.
+
+                    Here are all of the sources that you should refer to
+                    for context:
+                    {'\n'.join(context)}
+                """.strip(),
+            },
+            {
+                "role": "user",
+                "content": f"""
```
Comment (jxnl) [inline docs/prompting/thought_generation/chain_of_thought_zero_shot/thread_of_thought.md:53]: why f"{query}" its the same as `query`

File: docs/prompting/zero_shot/re2.md [hunk 6]
```diff
@@ -27,7 +27,7 @@ client = instructor.from_openai(OpenAI())
 
 
 class Solution(BaseModel):
-    final_answer: int = Field(..., description="This is the final answer")
+    final_answer: int = Field(..., description="Final Answer")
```
Comment (jxnl) [inline docs/prompting/zero_shot/re2.md:30]: why would we need the description final answer if the variable name is final answer

File: docs/prompting/zero_shot/self_ask.md [hunk 7]
```diff
@@ -1,7 +1,92 @@
...
+        ],
+    )
+
+
+if __name__ == "__main__":
+    query = """Who was president of the U.S.
+    when superconductivity was discovered?"""
+
+    response = generate_response(query)
+    print(response.follow_up_questions)
```
Comment (jxnl) [inline docs/prompting/zero_shot/self_ask.md:70]: just do `model_json_dumps(indent=2)`

**Review by jxnl — APPROVED** [review]

File: docs/prompting/thought_generation/chain_of_thought_zero_shot/tab_cot.md [hunk 8]
```diff
@@ -1,7 +1,122 @@
...
+    procedure: str = Field(
+        description="""Any intermediate computation
+        that was done in the reasoning process. Leave
+        empty if no computation is needed""",
+    )
+    result: str
+
+
+class Response(BaseModel):
+    reasoning: list[ReasoningStep] = Field(
```
Comment (jxnl) [inline docs/prompting/thought_generation/chain_of_thought_zero_shot/tab_cot.md:30]: @ivanleomk can you run some tests to check to make sure the reasoning stuff always show up before the correct_answer? we should also call out in the docs that 'correct_answer' is better than a generic variable like 'response' or 'content'

File: docs/prompting/thought_generation/chain_of_thought_zero_shot/tab_cot.md [hunk 9]
```diff
@@ -1,7 +1,122 @@
...
+            {
+                "role": "system",
+                "content": dedent(
+                    f"""
+                You are an expert Question Answering system. Make sure
+                to output your reasoning in structured reasoning steps
+                before generating a response to the user's query.
+
+
+                Context:
```
Comment (jxnl) [inline docs/prompting/thought_generation/chain_of_thought_zero_shot/tab_cot.md:50]: lets use xml,

**Review by jxnl — APPROVED** [review]

File: docs/prompting/thought_generation/chain_of_thought_zero_shot/tab_cot.md (hunk 8 above)
Comment (ivanleomk) [reply docs/prompting/thought_generation/chain_of_thought_zero_shot/tab_cot.md:30]: I verified that the reasoning shows up before the correct answer by using the `Partial`.

Output
```bash
reasoning=[PartialReasoningStep(step=1, subquestion='How many loaves of bread were sold in total?', procedure='93 + 39', result='132'), PartialReasoningStep(step=2, subquestion='What is the remaining number of loaves before accounting for returns?', procedure='200 - 132', result='68'), PartialReasoningStep(step=3, subquestion='How many loaves were left after accounting for the returned loaves?', procedure='68 + 6', result='74')] correct_answer=None
reasoning=[PartialReasoningStep(step=1, subquestion='How many loaves of bread were sold in total?', procedure='93 + 39', result='132'), PartialReasoningStep(step=2, subquestion='What is the remaining number of loaves before accounting for returns?', procedure='200 - 132', result='68'), PartialReasoningStep(step=3, subquestion='How many loaves were left after accounting for the returned loaves?', procedure='68 + 6', result='74')] correct_answer=None
reasoning=[PartialReasoningStep(step=1, subquestion='How many loaves of bread were sold in total?', procedure='93 + 39', result='132'), PartialReasoningStep(step=2, subquestion='What is the remaining number of loaves before accounting for returns?', procedure='200 - 132', result='68'), PartialReasoningStep(step=3, subquestion='How many loaves were left after accounting for the returned loaves?', procedure='68 + 6', result='74')] correct_answer=None
reasoning=[PartialReasoningStep(step=1, subquestion='How many loaves of bread were sold in total?', procedure='93 + 39', result='132'), PartialReasoningStep(step=2, subquestion='What is the remaining number of loaves before accounting for returns?', procedure='200 - 132', result='68'), PartialReasoningStep(step=3, subquestion='How many loaves were left after accounting for the returned loaves?', procedure='68 + 6', result='74')] correct_answer=74
reasoning=[PartialReasoningStep(step=1, subquestion='How many loaves of bread were sold in total?', procedure='93 + 39', result='132'), PartialReasoningStep(step=2, subquestion='What is the remaining number of loaves before accounting for returns?', procedure='200 - 132', result='68'), PartialReasoningStep(step=3, subquestion='How many loaves were left after accounting for the returned loaves?', procedure='68 + 6', result='74')] correct_answer=74
```


Code
```python
import instructor
from openai import OpenAI
from pydantic import BaseModel, Field
from textwrap import dedent


client = instructor.from_openai(OpenAI())


class ReasoningStep(BaseModel):
    step: int = Field(description="The step number")
    subquestion: str = Field(description="Subquestion to solve")
    procedure: str = Field(
        description="""Any intermediate computation
        that was done in the reasoning process. Leave
        empty if no computation is needed""",
    )
    result: str


class Response(BaseModel):
    reasoning: list[ReasoningStep] = Field(
        description="reasoning steps to derive answer",
    )
    correct_answer: int


def generate_structured_reasoning_response(query: str, context: str):
    response = client.chat.completions.create(
        model="gpt-4o",
        response_model=instructor.Partial[Response],
        messages=[
            {
                "role": "system",
                "content": dedent(
                    f"""
                You are an expert Question Answering system. Make sure
                to output your reasoning in structured reasoning steps
                before generating a response to the user's query.


                Context:
                {context}

                Query:
                {query}
                """
                ),
            },
        ],
        stream=True,
    )
    return response


if __name__ == "__main__":
    query = "How many loaves of bread did they have left?"
    context = """
    The bakers at the Beverly Hills Bakery baked
    200 loaves of bread on Monday morning. They
    sold 93 loaves in the morning and 39 loaves
    in the afternoon. A grocery store returned 6
    unsold loaves.
    """

    for response in generate_structured_reasoning_response(query, context):
        print(response)
```

File: docs/prompting/thought_generation/chain_of_thought_zero_shot/tab_cot.md (hunk 9 above)
Comment (ivanleomk) [reply docs/prompting/thought_generation/chain_of_thought_zero_shot/tab_cot.md:50]: Is this what you had in mind @jxnl 

```
def generate_structured_reasoning_response(query: str, context: str):
    response = client.chat.completions.create(
        model="gpt-4o",
        response_model=Response,
        messages=[
            {
                "role": "system",
                "content": dedent(
                    f"""
                <system>
                    <role>expert Question Answering system</role>
                    <instruction>Make sure to output your reasoning in structured reasoning steps before generating a response to the user's query.</instruction>
                </system>

                <context>
                    {context}
                </context>

                <query>
                    {query}
                </query>
                """
                ),
            },
        ],
    )
    return response
```

**Pull request merged**